src/backend/distributed/master/master_cache_table_shards.c

```diff
@@ -0,0 +1,183 @@
+ */
+
+#include "postgres.h"
+#include "funcapi.h"
+
+#include "distributed/master_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/worker_manager.h"
+#include "distributed/worker_protocol.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+
+static void DropShardsFromWorker(WorkerNode *workerNode, char *relationName,
+								 List *shardIntervalList);
+
+PG_FUNCTION_INFO_V1(master_expire_table_cache);
+
+
+/*
+ * master_expire_table_cache drops table's caches shards in all workers. The function
+ * expects a passed table to be a small distributed table meaning it has less than
+ * large_table_shard_count.
+ */
+Datum
+master_expire_table_cache(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *tableName = get_rel_name(relationId);
+	char *tableQualifiedName = quote_qualified_identifier(schemaName, tableName);
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *workerNodeList = NIL;
+	ListCell *workerNodeCell = NULL;
+	List *workerDroppedShardList = NIL;
+	ListCell *workerDroppedShardListCell = NULL;
+	int shardCount = 0;
+	List **placementListArray = NULL;
+	int shardIndex = 0;
+
+	if (!cacheEntry->isDistributedTable)
+	{
+		ereport(ERROR, (errmsg("Must be called on a distributed table")));
+	}
+
+	shardCount = cacheEntry->shardIntervalArrayLength;
+	if (shardCount == 0)
+	{
+		ereport(WARNING, (errmsg("Table has no shards, no action is taken")));
+		PG_RETURN_VOID();
+	}
+
+	if (shardCount >= LargeTableShardCount)
+	{
+		ereport(ERROR, (errmsg("Must be called on tables smaller than %d shards",
+							   LargeTableShardCount)));
+	}
+
+	placementListArray = palloc(shardCount * sizeof(List *));
+
+	for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+	{
+		ShardInterval *shardInterval = cacheEntry->sortedShardIntervalArray[shardIndex];
+		placementListArray[shardIndex] =
+			FinalizedShardPlacementList(shardInterval->shardId);
+	}
+
+	workerNodeList = WorkerNodeList();
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		List *currentWorkerDropList = NIL;
+
+		for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+		{
+			ShardInterval *shardInterval =
+				cacheEntry->sortedShardIntervalArray[shardIndex];
+			List *placementList = placementListArray[shardIndex];
+			ListCell *placementCell = NULL;
+			bool found = false;
+
+			foreach(placementCell, placementList)
+			{
+				ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(placementCell);
+
+				if (shardPlacement->nodePort == workerNode->workerPort &&
+					strncmp(shardPlacement->nodeName, workerNode->workerName,
+							WORKER_LENGTH) == 0)
+				{
+					found = true;
+					break;
+				}
```

Comment:
looks more difficult to follow,  what about placementCell == list_tail(placementList)